File: Andromeda.AvaloniaApp/ViewModels/Windows/InstallWindowViewModel.cs
using Andromeda.Core.FSharp;
using Avalonia.Controls;
using DynamicData;
using GogApi.DotNet.FSharp;
using ReactiveUI;
using ReactiveUI.Legacy;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Reactive;
using System.Text;
using System.Threading.Tasks;

using Andromeda.AvaloniaApp.Helpers;
using Andromeda.AvaloniaApp.ViewModels.Widgets;
using Andromeda.AvaloniaApp.Windows;

namespace Andromeda.AvaloniaApp.ViewModels.Windows
{
    public class InstallWindowViewModel : SubViewModelBase
    {
        private string gameSearchTerm = "";

        public string GameSearchTerm
        {
            get => this.gameSearchTerm;
            private set { this.RaiseAndSetIfChanged(ref this.gameSearchTerm, value); }
        }

        private ISourceList<Listing.ProductInfo> foundGames = new SourceList<Listing.ProductInfo>();

        public ISourceList<Listing.ProductInfo> FoundGames
        {
            get => this.foundGames;
            private set { this.RaiseAndSetIfChanged(ref this.foundGames, value); }
        }

        public ReactiveCommand<Unit, Unit> SearchGameCommand { get; }

        public InstallWindowViewModel(Control control, ViewModelBase parent) : base(control, parent)
        {
            SearchGameCommand = ReactiveCommand.Create(SearchGame);
        }

        void SearchGame()
        {
            var result = Andromeda.Core.FSharp.Games.getAvailableGamesForSearch(this.AppData, GameSearchTerm);
            this.SetAppData(result.Item2);
            if (result.Item1 != null)
            {
                var list = result.Item1.Value.ToList();
                //this.FoundGames.AddRange(list);
                if (list.Count > 0)
                {
                    var game = list.First();
                    var installersTuple = Games.getAvailableInstallersForOs(this.AppData, game.id);
                    this.SetAppData(installersTuple.Item2);
                    var list2 = installersTuple.Item1.ToList();
                    if (list2.Count() > 0)
                    {
                        var installerInfo = list2.First();

                        var downloadWidgetVM = this.Parent.GetChildrenOfType<DownloadWidgetViewModel>().FirstOrDefault();
                        Debug.Assert(downloadWidgetVM != null);
                        downloadWidgetVM.AddDownload(new InstallationInfos(game.title, installerInfo));
                    }
                } else {
                    Logger.LogWarning("Found no matching game for search term!");
                }
            }
        }
    }
}

File: Andromeda.AvaloniaApp.FSharp/ViewModels/InstallWindowViewModel.fs
namespace Andromeda.AvaloniaApp.FSharp.ViewModels

open Andromeda.Core.FSharp.Games
open Andromeda.Core.FSharp.Installed
open ReactiveUI
open DynamicData
open GogApi.DotNet.FSharp.Listing
open System.Diagnostics
open System.Linq
open System.Reactive
open Avalonia.Controls

open Andromeda.AvaloniaApp.FSharp.Helpers

type InstallWindowViewModel(control, parent) as this =
    inherit SubViewModelBase(control, parent)

    let gameSearchTerm = ""
    let foundGames = new SourceList<ProductInfo>()

    member private this.GameSearchTerm
        with get() = gameSearchTerm
        and set value = this.RaiseAndSetIfChanged(ref gameSearchTerm, value) |> ignore

    member private this.FoundGames
        with get() = foundGames
        and set value = this.RaiseAndSetIfChanged(ref foundGames, value) |> ignore

    member this.SearchGame() =
        let (games, appData) = getAvailableGamesForSearch this.AppData this.GameSearchTerm;
        appData |> this.SetAppData
        match games with
        | Some games when List.isEmpty games |> not ->
            //this.FoundGames.AddRange(list);
            let game = games.First()
            let (installers, appData) = getAvailableInstallersForOs this.AppData game.id
            appData |> this.SetAppData
            match List.isEmpty installers |> not with
            | true ->
                let installerInfo = installers.First();
                // TODO: refactor
                let downloadWidgetVM::_ =
                    this.GetRootViewModel().GetChildrenOfType<DownloadWidgetViewModel>()
                downloadWidgetVM.AddDownload(InstallationInfos(game.title, installerInfo))
            | false ->
                let message = "No installer for game found!"
                message |> Logger.LogWarning
                message |> this.GetRootViewModel().AddNotification
        | Some _ | None ->
            let message = "Found no matching game to install."
            message |> Logger.LogWarning
            message |> this.GetRootViewModel().AddNotification
        (this.Control :?> Window).Close()

    member val SearchGameCommand: ReactiveCommand<Unit, unit> = ReactiveCommand.Create<unit>(this.SearchGame)
